libr/include/r_util/r_str.h

```diff
@@ -19,6 +19,52 @@ typedef enum {
 	R_STRING_ENC_GUESS = 'g',
 } RStrEnc;
 
+// r_strf API can be used for short strings that are needed for a short period
+// of time. They are allocated on the stack as needed, by making sure the total
+// size of allocated strings does not become too big. If that is possible (e.g.
+// the input is not bounded and/or comes from the user) you should use the
+// malloc as usual.
+//
+// Use r_strf when you just need one string at a time. NOTE: Every time you use
+// r_strf, the previously created string can be replaced with the new one, so
+// make sure to use it only when you are done with the previous string
+//
+// Use r_strf_cont when you need to use multiple strings at a time (e.g.
+// multiple strings passed to the same function). NOTE: each string allocated
+// with r_strf_cont allocates other space on the stack, so be aware and use it
+// only for known short strings and not when inside loops as you risk of
+// allocating too much stuff.
+struct r_strf_t {
+	char *buf;
+	char *old_buf;
+	size_t sz;
+	size_t old_sz;
+	size_t total_sz;
+	int tmp;
+};
+
+#define R_STRF_FRAME r_strf_frame__
+#define r_strf_frame() struct r_strf_t R_STRF_FRAME = { NULL, NULL, 0, 0, 0 }
+#define r_strf(fmt, ...) (						\
+	R_STRF_FRAME.buf = R_STRF_FRAME.buf ? R_STRF_FRAME.buf : (R_STRF_FRAME.sz = 32, R_STRF_FRAME.old_sz = 32, R_STRF_FRAME.total_sz += 32, alloca (32)), \
+	R_STRF_FRAME.tmp = snprintf (R_STRF_FRAME.buf, R_STRF_FRAME.sz, fmt, __VA_ARGS__),                                                                   \
+	R_STRF_FRAME.sz = R_STRF_FRAME.tmp >= R_STRF_FRAME.sz ? R_STRF_FRAME.tmp * 2 : R_STRF_FRAME.sz,                                                      \
+	R_STRF_FRAME.total_sz + R_STRF_FRAME.sz >= 1024 * 5 ? r_assert_log (R_LOGLVL_FATAL, "r_strf used for long strings") : NULL,                          \
+	R_STRF_FRAME.buf = R_STRF_FRAME.sz != R_STRF_FRAME.old_sz ? (R_STRF_FRAME.total_sz += R_STRF_FRAME.sz, alloca (R_STRF_FRAME.sz)) : R_STRF_FRAME.buf, \
+	R_STRF_FRAME.tmp = R_STRF_FRAME.sz != R_STRF_FRAME.old_sz ? snprintf (R_STRF_FRAME.buf, R_STRF_FRAME.sz, fmt, __VA_ARGS__) : R_STRF_FRAME.tmp,       \
+	R_STRF_FRAME.old_sz = R_STRF_FRAME.sz,                                                                                                               \
+	R_STRF_FRAME.buf)
+
+#define r_strf_cont(fmt, ...) (                    \
```

Comment:
maybe `r_str_multi` instead of `r_str_cont`, but we can agree on a name :)

Reply:
this is my r_strf_rewind()

Reply:
also we can probably add another call r_strf_save() to save the current position and reset the buffer accordngly. this way it will be possible to define strings before entering a loop, save the position, use temporal variables inside the loop and then rewind to the saved index.

```
char *a = r_strf("Safe");
r_strf_save()
while(true) {
  char *b = r_strf("World");
  printf ("%s %s\n", a, b);
  r_strf_rewind();
}
```

Reply:
It is not your `r_strf_rewind`. Also, I don't see reasons to have two new functions, as  what you want is already possible as you can see in the tests I have provided (in particular see https://github.com/radareorg/radare2/pull/16259/files#diff-27544d81db79053eab8983dd74906e41R65 , where I first use `r_strf_cont` and then I call `r_strf` as many times as I want and the previously allocated strings with `r_strf_cont` are still there).